## 예외

- 예외는 error의 일종이며 프로그램이 수행 시 또는 컴파일시에 프로그램을 불능상태를 만들어버린다
  - 예외처리는 이러한 예외가 발생할 것을 미리 예측해서 이를 소스상에서 제어하고 처리하는 작업니다.



#### Java의 예외 

```Java
try {
    // 시도할 코드들
} catch (<예외 클래스 1> 변수명) {
    // 예외가 발생할 경우 해당 예외를 처리할 코드
} catch (<예외 클래스 2> 변수명) {
    // 예외가 발생할 경우 해당 예외를 처리할 코드
    // catch는 하나 이상 올 수 있다
} catch (Exception e) { 
    // Exception : 모든 예외의 최상위 부모. 특정하기 어렵거나 모든 예외를 잡고 싶을 때 사용
    // 예외가 발생할 경우 해당 예외를 처리할 코드
} finally {
    // 예외 발생 여부와 상관없이 항상 실행되는 코드
    // finally는 생략 가능하다
}
```

- 단 부모 예외 클래스가 자식 예외 클래스보다 먼저 나오면 안된다



### 예외가 발생했을 때 진행 순서

1. try 블록의 실행이 중단됨
2. catch 블록 중에 발생한 예외를 처리할 수 있는지 찾음
   - 위에서부터 하나씩 평가
3. 예외를 처리할 수 있는 catch 블록이 있다면. 해당 catch 블록안의 코드들이 실행 후 finally 블록 실행. try블록 이후의 코드들이 실행됨
4.  예외를 처할 수 있는 catch 블록이 없다면
   - finally블록을 실행 후 한 단계 높은 try 블록으로 전달



### custom 예외 만들기

- Exception을 상속받는 클래스를 상속받는다 (Exception을 상속받아도 된다)
- 자식 클래스들의 생성자에서 반드시 `super()`를 호출해야 한다
- 호출하지 않을 경우 message, cause가 저장되지 않는다
- message는 getMessage()로 확인 가능하다
- cause는 getCause()로 확인 가능하다

```java
public final class UserNotFoundException extends RuntimeException {
    public UserNotFoundException() {
        super();
    }
    
    public UserNotFoundException(String message) {
        super(message);
    }
    
    public UserNOtFolundException(String message, Throwable cause) {
        super(message, cause);
    }
}
```

